File: test_backend_connection.js
#!/usr/bin/env node
/**
 * Script zum Testen der Backend-Frontend Verbindung auf Render
 * Kann im Browser (Developer Console) oder mit Node.js ausgeführt werden
 */

// KONFIGURATION - Passen Sie diese URLs an
const BACKEND_URL = "https://poker-tracker-backend-3x39.onrender.com";
const FRONTEND_URL = "https://poker-tracker-frontend.onrender.com"; // Anpassen!

// Hilfsfunktionen für Console-Ausgabe
const log = {
    success: (msg) => console.log(`✓ ${msg}`),
    error: (msg) => console.error(`✗ ${msg}`),
    info: (msg) => console.log(`ℹ ${msg}`),
    warning: (msg) => console.warn(`⚠ ${msg}`),
    header: (msg) => {
        console.log('\n' + '='.repeat(60));
        console.log(msg);
        console.log('='.repeat(60) + '\n');
    }
};

// Test 1: Backend Erreichbarkeit
async function testBackendHealth() {
    log.header('TEST 1: Backend Erreichbarkeit');

    try {
        const response = await fetch(`${BACKEND_URL}/api/stats`);

        if (response.ok) {
            log.success(`Backend ist erreichbar unter ${BACKEND_URL}`);
            log.info(`Response Status: ${response.status}`);
            return true;
        } else {
            log.warning(`Backend antwortet mit Status ${response.status}`);
            return false;
        }
    } catch (error) {
        log.error(`Backend nicht erreichbar: ${error.message}`);
        return false;
    }
}

// Test 2: Test-Session erstellen
async function createTestSession() {
    log.header('TEST 2: Test-Session erstellen');

    const timestamp = new Date().toTimeString().split(' ')[0].replace(/:/g, '');

    const testData = {
        players: [
            {
                name: `TestPlayer1_${timestamp}`,
                buy_in: 100,
                rebuys: 1,
                endChips: 150
            },
            {
                name: `TestPlayer2_${timestamp}`,
                buy_in: 100,
                rebuys: 0,
                endChips: 75
            },
            {
                name: `TestPlayer3_${timestamp}`,
                buy_in: 100,
                rebuys: 2,
                endChips: 175
            }
        ]
    };

    log.info('Sende folgende Daten:');
    console.log(JSON.stringify(testData, null, 2));

    try {
        const response = await fetch(`${BACKEND_URL}/api/sessions`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(testData)
        });

        if (response.ok) {
            const data = await response.json();
            const sessionId = data.id;
            log.success(`Session erfolgreich erstellt! ID: ${sessionId}`);
            log.info(`Response: ${JSON.stringify(data, null, 2)}`);
            return sessionId;
        } else {
            log.error(`Fehler beim Erstellen: Status ${response.status}`);
            const text = await response.text();
            log.error(`Response: ${text}`);
            return null;
        }
    } catch (error) {
        log.error(`Anfrage fehlgeschlagen: ${error.message}`);
        return null;
    }
}

// Test 3: Session verifizieren
async function verifySession(sessionId) {
    log.header(`TEST 3: Session ${sessionId} verifizieren`);

    try {
        const response = await fetch(`${BACKEND_URL}/api/sessions/${sessionId}`);

        if (response.ok) {
            const data = await response.json();
            log.success('Session wurde gefunden!');
            log.info('Session Details:');
            console.log(JSON.stringify(data, null, 2));

            if (data.players && data.players.length === 3) {
                log.success('Alle 3 Spieler wurden gespeichert');

                data.players.forEach(player => {
                    log.info(`Spieler: ${player.name}, Net: ${player.net}, End Chips: ${player.end_chip}`);
                });

                return true;
            } else {
                log.warning('Spieleranzahl stimmt nicht überein');
                return false;
            }
        } else {
            log.error(`Session nicht gefunden: Status ${response.status}`);
            return false;
        }
    } catch (error) {
        log.error(`Verifizierung fehlgeschlagen: ${error.message}`);
        return false;
    }
}

// Test 4: History Endpoint
async function testHistoryEndpoint() {
    log.header('TEST 4: History Endpoint testen');

    try {
        const response = await fetch(`${BACKEND_URL}/api/history`);

        if (response.ok) {
            const sessions = await response.json();
            log.success(`History abgerufen: ${sessions.length} Sessions gefunden`);

            if (sessions.length > 0) {
                log.info('Letzte 3 Sessions:');
                sessions.slice(0, 3).forEach(session => {
                    console.log(`  - ID: ${session.id}, Datum: ${session.date}, ` +
                              `Gewinner: ${session.winner}, Spieler: ${session.player_count}`);
                });
            }

            return true;
        } else {
            log.error(`History Fehler: Status ${response.status}`);
            return false;
        }
    } catch (error) {
        log.error(`History abruf fehlgeschlagen: ${error.message}`);
        return false;
    }
}

// Test 5: Odds Calculator
async function testOddsCalculator() {
    log.header('TEST 5: Odds Calculator testen');

    const testData = {
        player_hand: ['Ah', 'Kh'],
        community_cards: ['Qh', 'Jh', '2c'],
        opponent_count: 2
    };

    log.info('Teste Odds Calculation mit:');
    console.log(JSON.stringify(testData, null, 2));

    try {
        const response = await fetch(`${BACKEND_URL}/api/odds`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(testData)
        });

        if (response.ok) {
            const data = await response.json();
            log.success('Odds Calculator funktioniert!');
            log.info(`Win Rate: ${data.win_rate}%, Equity: ${data.equity}%`);
            return true;
        } else {
            log.error(`Odds Calculator Fehler: Status ${response.status}`);
            return false;
        }
    } catch (error) {
        log.error(`Odds Calculator fehlgeschlagen: ${error.message}`);
        return false;
    }
}

// Hauptfunktion
async function runAllTests() {
    console.log('\n' + '='.repeat(60));
    console.log('🎰 Poker Tracker - Backend Connection Test');
    console.log('='.repeat(60) + '\n');

    log.info(`Backend URL: ${BACKEND_URL}`);
    log.info(`Frontend URL: ${FRONTEND_URL}`);
    log.info(`Zeitpunkt: ${new Date().toLocaleString()}\n`);

    const results = {};

    // Test 1: Backend Erreichbarkeit
    results.backend_health = await testBackendHealth();
    if (!results.backend_health) {
        log.error('\n❌ Backend nicht erreichbar - Tests abgebrochen');
        return;
    }

    // Test 2: Session erstellen
    const sessionId = await createTestSession();
    results.create_session = sessionId !== null;

    // Test 3: Session verifizieren
    if (sessionId) {
        results.verify_session = await verifySession(sessionId);
    } else {
        results.verify_session = false;
        log.warning('Session-Verifizierung übersprungen (keine ID)');
    }

    // Test 4: History
    results.history = await testHistoryEndpoint();

    // Test 5: Odds Calculator
    results.odds = await testOddsCalculator();

    // Zusammenfassung
    log.header('📊 TEST ZUSAMMENFASSUNG');

    const totalTests = Object.keys(results).length;
    const passedTests = Object.values(results).filter(v => v).length;

    Object.entries(results).forEach(([testName, passed]) => {
        const status = passed ? '✓ PASS' : '✗ FAIL';
        console.log(`${status} - ${testName}`);
    });

    console.log(`\nErgebnis: ${passedTests}/${totalTests} Tests bestanden\n`);

    if (passedTests === totalTests) {
        log.success('🎉 Alle Tests erfolgreich! Backend ist voll funktionsfähig!');

        if (sessionId) {
            console.log(`\nTest-Session ID: ${sessionId}`);
            log.info('Sie können diese Session in der History ansehen:');
            log.info(`  ${BACKEND_URL}/api/sessions/${sessionId}`);
        }
    } else {
        log.error('⚠️  Einige Tests sind fehlgeschlagen. Bitte Logs überprüfen.');
    }

    return results;
}

// Exportiere für Browser oder Node.js
if (typeof module !== 'undefined' && module.exports) {
    // Node.js
    module.exports = { runAllTests, testBackendHealth, createTestSession, verifySession };

    // Wenn direkt ausgeführt
    if (require.main === module) {
        runAllTests().catch(console.error);
    }
} else {
    // Browser - Stelle Funktionen global zur Verfügung
    window.pokerTrackerTest = {
        runAllTests,
        testBackendHealth,
        createTestSession,
        verifySession,
        testHistoryEndpoint,
        testOddsCalculator
    };

    console.log('🎰 Poker Tracker Test geladen!');
    console.log('Führe aus: pokerTrackerTest.runAllTests()');
}

File: test_backend_connection.py
#!/usr/bin/env python3
"""
Script zum Testen der Backend-Frontend Verbindung auf Render
Dieses Script testet:
1. Backend-Erreichbarkeit
2. Erstellen einer Test-Session
3. Abrufen der erstellten Session
4. Überprüfung der Daten
"""

import requests
import json
import sys
from datetime import datetime

# KONFIGURATION - Passen Sie diese URLs an
BACKEND_URL = "https://poker-tracker-backend-3x39.onrender.com"
FRONTEND_URL = "https://poker-tracker-frontend.onrender.com"  # Anpassen!

# Farben für Terminal-Ausgabe
class Colors:
    GREEN = '\033[92m'
    RED = '\033[91m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    END = '\033[0m'
    BOLD = '\033[1m'

def print_success(msg):
    print(f"{Colors.GREEN}✓ {msg}{Colors.END}")

def print_error(msg):
    print(f"{Colors.RED}✗ {msg}{Colors.END}")

def print_info(msg):
    print(f"{Colors.BLUE}ℹ {msg}{Colors.END}")

def print_warning(msg):
    print(f"{Colors.YELLOW}⚠ {msg}{Colors.END}")

def print_header(msg):
    print(f"\n{Colors.BOLD}{Colors.BLUE}{'='*60}{Colors.END}")
    print(f"{Colors.BOLD}{Colors.BLUE}{msg}{Colors.END}")
    print(f"{Colors.BOLD}{Colors.BLUE}{'='*60}{Colors.END}\n")

def test_backend_health():
    """Test ob das Backend erreichbar ist"""
    print_header("TEST 1: Backend Erreichbarkeit")

    try:
        # Versuche die Stats-Endpoint zu erreichen
        response = requests.get(f"{BACKEND_URL}/api/stats", timeout=10)

        if response.status_code == 200:
            print_success(f"Backend ist erreichbar unter {BACKEND_URL}")
            print_info(f"Response Status: {response.status_code}")
            return True
        else:
            print_warning(f"Backend antwortet mit Status {response.status_code}")
            return False

    except requests.exceptions.RequestException as e:
        print_error(f"Backend nicht erreichbar: {str(e)}")
        return False

def create_test_session():
    """Erstelle eine Test-Session im Backend"""
    print_header("TEST 2: Test-Session erstellen")

    test_data = {
        "players": [
            {
                "name": f"TestPlayer1_{datetime.now().strftime('%H%M%S')}",
                "buy_in": 100,
                "rebuys": 1,
                "endChips": 150
            },
            {
                "name": f"TestPlayer2_{datetime.now().strftime('%H%M%S')}",
                "buy_in": 100,
                "rebuys": 0,
                "endChips": 75
            },
            {
                "name": f"TestPlayer3_{datetime.now().strftime('%H%M%S')}",
                "buy_in": 100,
                "rebuys": 2,
                "endChips": 175
            }
        ]
    }

    print_info("Sende folgende Daten:")
    print(json.dumps(test_data, indent=2))

    try:
        response = requests.post(
            f"{BACKEND_URL}/api/sessions",
            json=test_data,
            headers={"Content-Type": "application/json"},
            timeout=10
        )

        if response.status_code == 200:
            data = response.json()
            session_id = data.get('id')
            print_success(f"Session erfolgreich erstellt! ID: {session_id}")
            print_info(f"Response: {json.dumps(data, indent=2)}")
            return session_id
        else:
            print_error(f"Fehler beim Erstellen: Status {response.status_code}")
            print_error(f"Response: {response.text}")
            return None

    except requests.exceptions.RequestException as e:
        print_error(f"Anfrage fehlgeschlagen: {str(e)}")
        return None

def verify_session(session_id):
    """Überprüfe ob die Session korrekt gespeichert wurde"""
    print_header(f"TEST 3: Session {session_id} verifizieren")

    try:
        # 1. Hole spezifische Session
        response = requests.get(
            f"{BACKEND_URL}/api/sessions/{session_id}",
            timeout=10
        )

        if response.status_code == 200:
            data = response.json()
            print_success("Session wurde gefunden!")
            print_info("Session Details:")
            print(json.dumps(data, indent=2))

            # Validiere Daten
            if 'players' in data and len(data['players']) == 3:
                print_success("Alle 3 Spieler wurden gespeichert")

                # Überprüfe Berechnungen
                for player in data['players']:
                    expected_net = player['end_chip'] - (player['end_chip'] - player['net'])
                    print_info(f"Spieler: {player['name']}, Net: {player['net']}, End Chips: {player['end_chip']}")

                return True
            else:
                print_warning("Spieleranzahl stimmt nicht überein")
                return False
        else:
            print_error(f"Session nicht gefunden: Status {response.status_code}")
            return False

    except requests.exceptions.RequestException as e:
        print_error(f"Verifizierung fehlgeschlagen: {str(e)}")
        return False

def test_history_endpoint():
    """Test ob die History alle Sessions enthält"""
    print_header("TEST 4: History Endpoint testen")

    try:
        response = requests.get(f"{BACKEND_URL}/api/history", timeout=10)

        if response.status_code == 200:
            sessions = response.json()
            print_success(f"History abgerufen: {len(sessions)} Sessions gefunden")

            if len(sessions) > 0:
                print_info("Letzte 3 Sessions:")
                for session in sessions[:3]:
                    print(f"  - ID: {session['id']}, Datum: {session['date']}, "
                          f"Gewinner: {session['winner']}, Spieler: {session['player_count']}")

            return True
        else:
            print_error(f"History Fehler: Status {response.status_code}")
            return False

    except requests.exceptions.RequestException as e:
        print_error(f"History abruf fehlgeschlagen: {str(e)}")
        return False

def test_cors_headers():
    """Teste CORS Headers vom Backend"""
    print_header("TEST 5: CORS-Konfiguration prüfen")

    try:
        response = requests.options(
            f"{BACKEND_URL}/api/sessions",
            headers={
                "Origin": FRONTEND_URL,
                "Access-Control-Request-Method": "POST",
                "Access-Control-Request-Headers": "Content-Type"
            },
            timeout=10
        )

        cors_headers = {
            "Access-Control-Allow-Origin": response.headers.get("Access-Control-Allow-Origin"),
            "Access-Control-Allow-Methods": response.headers.get("Access-Control-Allow-Methods"),
            "Access-Control-Allow-Headers": response.headers.get("Access-Control-Allow-Headers"),
        }

        print_info("CORS Headers:")
        for key, value in cors_headers.items():
            if value:
                print_success(f"  {key}: {value}")
            else:
                print_warning(f"  {key}: Nicht gesetzt")

        if cors_headers.get("Access-Control-Allow-Origin"):
            print_success("CORS ist aktiviert")
            return True
        else:
            print_warning("CORS Headers nicht optimal konfiguriert")
            return True  # Nicht kritisch, da CORS via Flask-CORS gesetzt wird

    except requests.exceptions.RequestException as e:
        print_warning(f"CORS-Test nicht möglich: {str(e)}")
        return True

def test_frontend_to_backend():
    """Test die Frontend-zu-Backend Weiterleitung"""
    print_header("TEST 6: Frontend-Weiterleitung (Optional)")

    print_info("Frontend URL: " + FRONTEND_URL)
    print_warning("Hinweis: Dieser Test kann nur funktionieren, wenn Frontend deployed ist")

    try:
        # Versuche über Frontend zu routen
        response = requests.get(f"{FRONTEND_URL}/api/stats", timeout=10)

        if response.status_code == 200:
            print_success("Frontend leitet korrekt zum Backend weiter!")
            return True
        else:
            print_warning(f"Frontend-Weiterleitung Status: {response.status_code}")
            print_info("Dies ist normal wenn Frontend noch nicht deployed ist")
            return True

    except requests.exceptions.RequestException as e:
        print_warning(f"Frontend-Test fehlgeschlagen: {str(e)}")
        print_info("Frontend könnte noch nicht deployed sein - Backend funktioniert aber!")
        return True

def main():
    print(f"\n{Colors.BOLD}{'='*60}")
    print("🎰 Poker Tracker - Backend Connection Test")
    print(f"{'='*60}{Colors.END}\n")

    print_info(f"Backend URL: {BACKEND_URL}")
    print_info(f"Frontend URL: {FRONTEND_URL}")
    print_info(f"Zeitpunkt: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")

    results = {}

    # Test 1: Backend Erreichbarkeit
    results['backend_health'] = test_backend_health()
    if not results['backend_health']:
        print_error("\n❌ Backend nicht erreichbar - Tests abgebrochen")
        sys.exit(1)

    # Test 2: Session erstellen
    session_id = create_test_session()
    results['create_session'] = session_id is not None

    # Test 3: Session verifizieren
    if session_id:
        results['verify_session'] = verify_session(session_id)
    else:
        results['verify_session'] = False
        print_warning("Session-Verifizierung übersprungen (keine ID)")

    # Test 4: History
    results['history'] = test_history_endpoint()

    # Test 5: CORS
    results['cors'] = test_cors_headers()

    # Test 6: Frontend Weiterleitung
    results['frontend'] = test_frontend_to_backend()

    # Zusammenfassung
    print_header("📊 TEST ZUSAMMENFASSUNG")

    total_tests = len(results)
    passed_tests = sum(1 for v in results.values() if v)

    for test_name, passed in results.items():
        status = "✓ PASS" if passed else "✗ FAIL"
        color = Colors.GREEN if passed else Colors.RED
        print(f"{color}{status}{Colors.END} - {test_name}")

    print(f"\n{Colors.BOLD}Ergebnis: {passed_tests}/{total_tests} Tests bestanden{Colors.END}\n")

    if passed_tests == total_tests:
        print_success("🎉 Alle Tests erfolgreich! Backend ist voll funktionsfähig!")

        if session_id:
            print(f"\n{Colors.BOLD}Test-Session ID: {session_id}{Colors.END}")
            print_info("Sie können diese Session in der History ansehen:")
            print_info(f"  {BACKEND_URL}/api/sessions/{session_id}")
    else:
        print_error("⚠️  Einige Tests sind fehlgeschlagen. Bitte Logs überprüfen.")
        sys.exit(1)

if __name__ == "__main__":
    main()

